webapi/Services/MemoryMigration/ChatMemoryMigrationService.cs

```diff
@@ -0,0 +1,146 @@
+        this._globalIndex = documentMemoryOptions.Value.GlobalDocumentCollectionName;
+        var kernel = provider.GetMigrationKernel();
+        this._memory = kernel.Memory;
+    }
+
+    ///<inheritdoc/>
+    public async Task MigrateAsync(CancellationToken cancellationToken = default)
+    {
+        try
+        {
+            await this.InternalMigrateAsync(cancellationToken).ConfigureAwait(false);
+        }
+        catch (Exception exception) when (!exception.IsCriticalException())
+        {
+            this._logger.LogError(exception, "Error migrating chat memories");
+        }
+    }
+
+    private async Task InternalMigrateAsync(CancellationToken cancellationToken = default)
+    {
+        var collectionNames = (await this._memory.GetCollectionsAsync(cancellationToken).ConfigureAwait(false)).ToHashSet(StringComparer.OrdinalIgnoreCase);
+
+        var tokenMemory = await GetTokenMemory(cancellationToken).ConfigureAwait(false);
+        if (tokenMemory != null)
+        {
+            //  Create memory token already exists
+            return;
+        }
+
+        //  Create memory token
+        var token = Guid.NewGuid().ToString();
+        await SetTokenMemory(token, cancellationToken).ConfigureAwait(false);
+
+        await RemoveMemorySourcesAsync().ConfigureAwait(false);
+
+        bool needsZombie = true;
+        // Extract and store memories, using the original id to avoid duplication should a retry be required.
+        await foreach ((string chatId, string memoryName, string memoryId, string memoryText) in QueryMemoriesAsync())
+        {
+            await this._memoryClient.StoreMemoryAsync(this._promptOptions.MemoryIndexName, chatId, memoryName, memoryId, memoryText, cancellationToken);
+            needsZombie = false;
+        }
+
+        if (needsZombie)
+        {
+            await this._memoryClient.StoreMemoryAsync(this._promptOptions.MemoryIndexName, Guid.Empty.ToString(), "zombie", Guid.NewGuid().ToString(), "Initialized", cancellationToken);
+        }
+
+        await SetTokenMemory(ChatMigrationMonitor.MigrationCompletionToken, cancellationToken).ConfigureAwait(false);
+
+        // Inline function to extract all memories for a given chat and memory type.
+        async IAsyncEnumerable<(string chatId, string memoryName, string memoryId, string memoryText)> QueryMemoriesAsync()
+        {
+            var chats = await this._chatSessionRepository.GetAllChatsAsync().ConfigureAwait(false);
+            foreach (var chat in chats)
+            {
+                foreach (var memoryType in this._promptOptions.MemoryMap.Keys)
+                {
+                    var indexName = $"{chat.Id}-{memoryType}";
+                    if (collectionNames.Contains(indexName))
+                    {
+                        var memories = await this._memory.SearchAsync(indexName, "*", limit: 10000, minRelevanceScore: 0, withEmbeddings: false, cancellationToken).ToArrayAsync(cancellationToken);
```

Comment:
Is 1000 enough?

Reply:
We don't know what the customer may have, but 10000 is the maximum allowed value (anything higher than that results in a fail-fast exception).